File: Napakalaki Ruby/lib/cultist_player.rb
# By: Miguel Robles Urquiza
#     Jesús Sánchez de Lechina Tejada

require_relative "player"

module NapakalakiGame

  class CultistPlayer < Player
    
    @@totalCultistPlayers = 0
    
    def initialize(p,c)
      @name = p.getName
      @level = p.getLevels
      @dead = p.isDead
      @canISteal = p.canISteal
      @hiddenTreasures = Array.new p.getHiddenTreasures
      @visibleTreasures = Array.new p.getVisibleTreasures
      @pendingBadConsequence = p.getPendingBadConsequence
      @enemy = p.getEnemy
      @myCultistCard = c
      @@totalCultistPlayers += 1


    end
    @Override
    def getCombatLevel
      (super * 1.7).truncate + @@totalCultistPlayers * @myCultistCard.getGainedLevels   
    end
    
    def getOponentLevel(m)
      m.getCombatLevelAgainstCultistPlayer      
    end
    
    def self.getTotalCultistPlayers 
      @@totalCultistPlayers
    end
    
    def shouldConvert
      false
    end
    
    def giveMeATreasure
      indice = rand(@visibleTreasures.size)
      tr = @visibleTreasures[indice]
      @visibleTreasures.delete_at(indice)
      tr
    end
    
    def canYouGiveMeATreasure
      true if (@visibleTreasures.size() > 0)
      false
    end
  end
end

File: Napakalaki Java/src/NapakalakiGame/CultistPlayer.java
/**
 *
 * @author Jesús Sánchez de Lechina Tejada & Miguel Ángel Robles Urquiza
 */
package NapakalakiGame;


public class CultistPlayer extends Player {
    public static int totalCultistPlayers = 0;
    public Cultist myCultistCard;
    
    public CultistPlayer(Player p, Cultist c){
        super(p.getName());
        level = p.getLevels();
        dead = p.isDead();
        canISteal = p.canISteal();
        hiddenTreasures = p.getHiddenTreasures();
        visibleTreasures = p.getVisibleTreasures();
        pendingBadConsequence = p.getPendingBadConsequence();
        enemy = p.getEnemy();
        myCultistCard = c;
        totalCultistPlayers += 1;
    }
    
    @Override
    public int getCombatLevel() {
        return (int) (Math.round(super.getCombatLevel() * 1.7) + totalCultistPlayers * myCultistCard.getGainedLevels());
    }
    
    protected int getOponentLevel(Monster m) {
        return m.getCombatLevelAgainstCultistPlayer();
    }
    
    protected boolean shouldConvert() {
        return false;
    }
    
    private Treasure giveMeATreasure(){
        int indice = (int) (Math.random() % visibleTreasures.size());
        return visibleTreasures.remove(indice);
    }
    
    private boolean canYouGiveMeATreasure(){
        return visibleTreasures.size() > 0;
    }
    
    public static int getTotalCultistPlayers() {
        return totalCultistPlayers;
    }
}
